### ISSUE TITLE
Separate Builders into internal and external to the build system

### ISSUE THREAD

**Issue by angerman** [snowleopard/shaking-up-ghc#124]
As #26 illustrates, we might have builders that are external to the build system, and the build system, which the build *can not* produce and we have to rely on them to exist. I hence propose that we allow
Builders to be annotated with `external` or `internal`, and only call `need` on those that are `internal`. If someone (autoconf) says an external builder exists, let's just assume that is true.

**Comment by snowleopard** [comment]
I agree, Builders are different and we should treat them differently.

Also at the moment we don't list all builders we use, for example, `configure`, `make`, `patch` and some others are called simply via `cmd`. We probably want a more coherent story. Let's discuss possible approaches here.

**Comment by snowleopard** [comment]
One simple first step is to change `needBuilder` by running it only on `specified` builders:
```haskell
needBuilder :: Bool -> Builder -> Action ()
needBuilder laxDependencies builder = whenM (specified builder) $ do
[...]
```
I think this makes good sense, so I pushed it.
/cc #123

**Comment by snowleopard** [comment]
I refactored `Builder.hs`, here is a summary:
* Function `builderProvenance :: Builder -> Maybe (Stage, Package)` sets up a correspondence between the builders we build and the `Stage`/`Package` combinations.
* Internal builders can now be characterised as `isInternal = isJust . builderProvenance`.
* `needBuilder` only needs internal builders. 
* Internal builders have been removed from `system.config`.
* Paths to internal builders are determined using `programPath`:

    ```haskell
    builderPath builder = case builderProvenance builder of
        Just (stage, pkg) -> return . fromJust $ programPath stage pkg
        Nothing -> do
            let builderKey = case builder of
    ```

Any comments/further improvements are welcome!
